staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -546,14 +556,30 @@ func (c unstructuredCreator) New(kind schema.GroupVersionKind) (runtime.Object,
 }
 
 type unstructuredDefaulter struct {
-	delegate runtime.ObjectDefaulter
+	delegate  runtime.ObjectDefaulter
+	defaulter *validate.SchemaValidator
 }
 
 func (d unstructuredDefaulter) Default(in runtime.Object) {
 	// Delegate for things other than Unstructured.
-	if _, ok := in.(runtime.Unstructured); !ok {
+	obj, ok := in.(runtime.Unstructured)
+	if !ok {
 		d.delegate.Default(in)
+		return
+	}
+
+	// apply defaults from the validator
+	if d.defaulter == nil {
+		return
 	}
+	result := d.defaulter.Validate(obj.UnstructuredContent())
+	if !result.IsValid() {
+		ns, _ := unstructured.NestedString(obj.UnstructuredContent(), "namespace")
+		name, _ := unstructured.NestedString(obj.UnstructuredContent(), "name")
+		utilruntime.HandleError(fmt.Errorf("validation error of CustomResource %s/%s: %v", ns, name, result.AsError()))
```

Comment:
Can you make it so that the only way this defaulter can be used is in the context of a decode, so that you can have a different Default() method and return an error?

The decision to fail open or closed on defaulting is kind of an important one.

Reply:
Failing open can result in clients breaking.

Reply:
error handling was an open question here as there is no error return type. Will investigate.